frame/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,301 @@
+	use frame_support::{pallet_prelude::*, traits::fungible::Mutate};
+	use frame_system::pallet_prelude::*;
+	use scale_info::TypeInfo;
+	use sp_core::keccak_256;
+	use sp_runtime::{
+		traits::{
+			AtLeast32BitUnsigned, CheckedAdd, CheckedMul, CheckedSub, Convert, Saturating, Verify,
+			Zero,
+		},
+		AccountId32, MultiSignature, Perbill,
+	};
+
+	use super::models::*;
+
+	#[derive(Encode, Decode, PartialEq, Copy, Clone, TypeInfo)]
+	pub struct Reward<Balance, BlockNumber> {
+		total: Balance,
+		claimed: Balance,
+		vesting_period: BlockNumber,
+	}
+
+	type RemoteAccountOf<T> = RemoteAccount<<T as Config>::RelayChainAccountId>;
+	type RewardOf<T> = Reward<<T as Config>::Balance, <T as frame_system::Config>::BlockNumber>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		Claimed {
+			remote_account: RemoteAccountOf<T>,
+			reward_account: T::AccountId,
+			amount: T::Balance,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AlreadyInitialized,
+		InvalidProof,
+		InvalidClaim,
+		NotAContributor,
+		NothingToClaim,
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		#[allow(missing_docs)]
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type Currency: Mutate<Self::AccountId, Balance = Self::Balance>;
+		type Balance: Default
+			+ Parameter
+			+ Codec
+			+ Copy
+			+ Ord
+			+ CheckedAdd
+			+ CheckedSub
+			+ CheckedMul
+			+ AtLeast32BitUnsigned
+			+ From<u128>
+			+ Into<u128>
+			+ Zero;
+		type AdminOrigin: EnsureOrigin<Self::Origin>;
+		type Convert: Convert<Self::BlockNumber, Self::Balance>;
+		type RelayChainAccountId: Parameter + Into<AccountId32> + Ord;
+		#[pallet::constant]
+		type InitialPayment: Get<Perbill>;
+		#[pallet::constant]
+		type VestingPartition: Get<Self::BlockNumber>;
+		#[pallet::constant]
+		type Prefix: Get<&'static [u8]>;
+	}
+
+	#[pallet::storage]
+	pub type Rewards<T: Config> =
+		StorageMap<_, Blake2_128Concat, RemoteAccountOf<T>, RewardOf<T>, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_rewards)]
+	pub type TotalRewards<T: Config> = StorageValue<_, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn claimed_rewards)]
+	pub type ClaimedRewards<T: Config> = StorageValue<_, T::Balance, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_contributors)]
+	pub type TotalContributors<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn vesting_block_start)]
+	pub type VestingBlockStart<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000)]
+		pub fn initialize(
+			origin: OriginFor<T>,
+			rewards: Vec<(RemoteAccountOf<T>, T::Balance, T::BlockNumber)>,
+		) -> DispatchResult {
```

Comment:
Most likely we will need to repeatedly call this, as the rewards vec is probably too large to fit into a single extrinsic call.

Reply:
As @seunlanlege said, we should use the genesis block.